www/codePush.ts

```diff
@@ -105,9 +112,122 @@ class CodePush implements CodePushCordovaPlugin {
+     * 
+     * @param syncCallback Optional callback to be called with the status of the sync operation.
+     *                     The callback will be called only once, and the possible statuses are defined by the SyncStatus enum. 
+     * @param syncOptions Optional SyncOptions parameter configuring the behavior of the sync operation.
+     * 
+     */
+    public sync(syncCallback?: SuccessCallback<any>, syncOptions?: SyncOptions): void {
+
+        /* No options were specified, use default */
+        if (!syncOptions) {
+            syncOptions = this.getDefaultSyncOptions();
+        }
+
+        /* Some options were specified */
+        if (syncOptions.mandatoryUpdateMessage) {
+            syncOptions.mandatoryUpdateContinueButtonLabel = syncOptions.mandatoryUpdateContinueButtonLabel || this.getDefaultSyncOptions().mandatoryUpdateContinueButtonLabel;
+            syncOptions.dialogTitle = syncOptions.dialogTitle || this.getDefaultSyncOptions().dialogTitle;
+        }
+        if (syncOptions.optionalUpdateMessage) {
+            syncOptions.optionalUpdateConfirmButtonLabel = syncOptions.optionalUpdateConfirmButtonLabel || this.getDefaultSyncOptions().optionalUpdateConfirmButtonLabel;
+            syncOptions.optionalUpdateCancelButtonLabel = syncOptions.optionalUpdateCancelButtonLabel || this.getDefaultSyncOptions().optionalUpdateCancelButtonLabel;
+            syncOptions.dialogTitle = syncOptions.dialogTitle || this.getDefaultSyncOptions().dialogTitle;
+        }
+
+        window.codePush.notifyApplicationReady();
+
+        var onError = (error: Error) => {
+            CodePushUtil.logError("An error occurred during sync.", error);
+            syncCallback && syncCallback(SyncStatus.ERROR);
+        };
+
+        var onApplySuccess = () => {
+            syncCallback && syncCallback(SyncStatus.APPLY_SUCCESS);
+        };
+
+        var onDownloadSuccess = (localPackage: ILocalPackage) => {
+            localPackage.apply(onApplySuccess, onError, syncOptions.rollbackTimeout);
+        };
+
+        var downloadAndInstallUpdate = (remotePackage: RemotePackage) => {
+            remotePackage.download(onDownloadSuccess, onError);
+        };
+
+        var onUpdate = (remotePackage: RemotePackage) => {
+            if (!remotePackage || (remotePackage.failedApply && syncOptions.ignoreFailedUpdates)) {
+                syncCallback && syncCallback(SyncStatus.UP_TO_DATE);
+            } else {
+                if (remotePackage.isMandatory && syncOptions.mandatoryUpdateMessage) {
+                    /* Alert user */
+                    navigator.notification.alert(syncOptions.mandatoryUpdateMessage, () => { downloadAndInstallUpdate(remotePackage); }, syncOptions.dialogTitle, syncOptions.mandatoryUpdateContinueButtonLabel);
+                } else if (!remotePackage.isMandatory && syncOptions && syncOptions.optionalUpdateMessage) {
```

Comment:
Nittish, but you don't need to check syncOptions exists here (and aren't doing that in the 'if' above anyway)